cirq/protocols/equal_up_to_global_phase.py

```diff
@@ -0,0 +1,107 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import Iterable
+import numbers
+from typing import Any, Union
+
+import numpy as np
+from typing_extensions import Protocol
+
+import cirq
+
+
+class SupportsEqualUpToGlobalPhase(Protocol):
+    """Object which can be compared for equality mod global phase."""
+
+    def _equal_up_to_global_phase_(self, other: Any, *,
+                                   atol: Union[int, float]) -> bool:
+        """Approximate comparator.
+
+        Types implementing this protocol define their own logic for comparison
+        with other types.
+
+        Args:
+            other: Target object for comparison of equality up to global phase.
+            atol: The minimum absolute tolerance. See np.isclose() documentation
+                  for details.
+
+        Returns:
+            True if objects are equal up to a global phase, False otherwise.
+            Returns NotImplemented when checking equality up to a global phase
+            is not implemented for given types.
+        """
+
+
+def equal_up_to_global_phase(val: Any,
+                             other: Any,
+                             *,
+                             atol: Union[int, float] = 1e-8) -> bool:
+    """Determine whether two objects are equal up to global phase.
+
+    If `val` implements SupportsEqualUpToGlobalPhase protocol then it is

```

Comment:
Document in terms of the method to implement, not the name of the protocol, as this is more direct information for the user.

Reply:
done